### PR TITLE
signalbackup-tools: 20220711 -> 20220810

### PR DESCRIPTION
**Pull request by malob**
###### Description of changes

Update package, and add patch to resolve compilation issues on Darwin systems.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by malob** [comment]
Result of `nixpkgs-review pr 185963` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>signalbackup-tools</li>
  </ul>
</details>

**Comment by malob** [comment]
Result of `nixpkgs-review pr 185963` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>signalbackup-tools</li>
  </ul>
</details>

File: pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix [hunk 1]
```diff
@@ -1,30 +1,31 @@
...
-    "-pedantic"
-    "-std=c++2a"
-    "-O3"
-    "-march=native"
+  patches = [
+    (fetchpatch {
+      name = "fix-platform-checks.patch";
+      url = "https://github.com/bepaald/signalbackup-tools/compare/20220810..a81baf25b6ba63da7d30d9a239e5b4bbc8d1ab4f.patch";
+      sha256 = "sha256-i7fuPBil8zB+V3wHHdcbmP79OZoTfG2ZpXPQ3m7X06c=";
+    })
```
Comment (winterqt) [inline pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix:19]: Please change this to only include [bepaald/signalbackup-tools@`47ff35e` (#73)](https://github.com/bepaald/signalbackup-tools/pull/73/commits/47ff35e85921e109c0d58d89f40fd3a7ef2d44ff) -- probably not worth including the gitignore changes 🙂

File: pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix [hunk 2]
```diff
@@ -1,30 +1,31 @@
...
   ];
-  buildPhase = ''
-    $CXX $buildFlags */*.cc *.cc -lcrypto -lsqlite3 -o signalbackup-tools
+
+  buildInputs = [ openssl sqlite ];
+  buildPhase = lib.optionalString stdenv.isDarwin ''
+    export CXXFLAGS="-Wall -Wextra -Wshadow -Wold-style-cast -Woverloaded-virtual -pedantic -O3"
+    export LDFLAGS="-Wall -Wextra -O3"
+  '' + ''
+    patchShebangs BUILDSCRIPT_MULTIPROC.bash44
```
Comment (winterqt) [inline pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix:27]: This should move to `postPatch`, since it's patching the file.

File: pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix [hunk 3]
```diff
@@ -1,30 +1,31 @@
...
+      sha256 = "sha256-i7fuPBil8zB+V3wHHdcbmP79OZoTfG2ZpXPQ3m7X06c=";
+    })
   ];
-  buildPhase = ''
-    $CXX $buildFlags */*.cc *.cc -lcrypto -lsqlite3 -o signalbackup-tools
+
+  buildInputs = [ openssl sqlite ];
+  buildPhase = lib.optionalString stdenv.isDarwin ''
+    export CXXFLAGS="-Wall -Wextra -Wshadow -Wold-style-cast -Woverloaded-virtual -pedantic -O3"
+    export LDFLAGS="-Wall -Wextra -O3"
```
Comment (winterqt) [inline pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix:25]: Why are these required on Darwin?

File: pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix [hunk 4]
```diff
@@ -1,30 +1,31 @@
...
+      name = "fix-platform-checks.patch";
+      url = "https://github.com/bepaald/signalbackup-tools/compare/20220810..a81baf25b6ba63da7d30d9a239e5b4bbc8d1ab4f.patch";
+      sha256 = "sha256-i7fuPBil8zB+V3wHHdcbmP79OZoTfG2ZpXPQ3m7X06c=";
+    })
   ];
-  buildPhase = ''
-    $CXX $buildFlags */*.cc *.cc -lcrypto -lsqlite3 -o signalbackup-tools
+
+  buildInputs = [ openssl sqlite ];
+  buildPhase = lib.optionalString stdenv.isDarwin ''
```
Comment (winterqt) [inline pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix:23]: Please `runHook preBuild` and `runHook postBuild` here. You should also fix your `installPhase` to do the same.

**Review by winterqt — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix (hunk 1 above)
Comment (malob) [reply pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix:19]: This patch is not based on that pull request. It was initially when I submitted the pull request but I quickly changed that (see this [diff](https://github.com/NixOS/nixpkgs/compare/578b3919bf1af70ed1198586101bb8ca5df153c1..af1f959f935fb9c34a1fd1135043cbdad302205d)) when the project's author implemented the change himself in a different way (https://github.com/bepaald/signalbackup-tools/commit/a81baf25b6ba63da7d30d9a239e5b4bbc8d1ab4f). 

This patch includes a range of commits since the patch for the relevant commit doesn't cleanly apply without including a couple other trivial commits the author has made since the most recent release.

File: pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix (hunk 2 above)
Comment (malob) [reply pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix:27]: Seems reasonable, will do.

File: pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix (hunk 3 above)
Comment (malob) [reply pkgs/applications/networking/instant-messengers/signalbackup-tools/default.nix:25]: In [`BUILDSCRIPT_MULTIPROC.bash44`](https://github.com/bepaald/signalbackup-tools/blob/a81baf25b6ba63da7d30d9a239e5b4bbc8d1ab4f/BUILDSCRIPT_MULTIPROC.bash44) these variables are defined as follows:

```bash
CXXFLAGS="${CXXFLAGS:--Wall -Wextra -Wshadow -Wold-style-cast -Woverloaded-virtual -pedantic -O3 -flto}"
LDFLAGS="${LDFLAGS:--Wall -Wextra -Wl,-z,now -Wl,--as-needed -O3 -s -flto=auto}"
```

Link time optimization currently doesn't work on for Darwin in `nixpkgs`, so the `-flto` flag in `CXXFLAGS` leads to linker errors. The linker on Darwin doesn't support the `-z` or `--as-needed` flags, and the `-s` is obsolete, so those all need to be removed from `LDFLAGS`.

**Comment by malob** [comment]
@SuperSandro2000, could you review/merge?

**Review by winterqt — APPROVED** [review]

**Comment by winterqt** [comment]
Since you requested Sandro's review, I'll wait on him in case he has anything to say; sorry about taking so long to get to this.

**Comment by malob** [comment]
> Since you requested Sandro's review, I'll wait on him in case he has anything to say; sorry about taking so long to get to this.

@winterqt, happy for you to merge, Sandro is just the dude I poke when I've had a PR open for a few days and I think it's ready to merge.

**Comment by winterqt** [comment]
I just want to make sure I'm not missing anything, doesn't hurt to have more eyes on something IMO. If he doesn't respond in a day or two I'll go ahead and merge.
